[WPE] Fix DRM initialization in nxp-imx6qp-sdb board

psaavedra opened the pull request:
#### 5a6315f2de686283ee9ffc3498e6b68107799566
<pre>
[WPE] Fix DRM initialization in nxp-imx6qp-sdb board
<a href="https://bugs.webkit.org/show_bug.cgi?id=268487">https://bugs.webkit.org/show_bug.cgi?id=268487</a>

Reviewed by Carlos Garcia Campos.

This patch is generic but fixes the DRM initialization for NXP imx6qp
Sabre Development boards:

* Removes the check for the valid connector in the chooseConnector().
  The encoder Id can be zero if the connector is not currently bound
  to an encoder or if the encoder+crtc is already used by another
  connector.

* Modifies the chooseCrtcForConnector() logic. We implement a more
  robust mechanism by checking the encoder-&gt;possible_crtcs in case of
  no active crtc was found.

Credits to Jani Hautakangas &lt;jani@igalia.com&gt;.

Related-To: <a href="https://github.com/Igalia/cog/pull/590">https://github.com/Igalia/cog/pull/590</a>
Related-To: <a href="https://github.com/WebKit/WebKit/pull/23654">https://github.com/WebKit/WebKit/pull/23654</a>
Related-To: <a href="https://github.com/WebKit/WebKit/pull/23657">https://github.com/WebKit/WebKit/pull/23657</a>

* Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp:
(chooseConnector):
(chooseCrtcForConnector):

Canonical link: <a href="https://commits.webkit.org/273898@main">https://commits.webkit.org/273898@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6c5d0b25a5d93b3c2f8aae88af326fd1d19cfde7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/37142 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/16046 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/14/builds/39440 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/39717 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/33159 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/18618 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/51/builds/13185 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/39717 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🧪 webkitperl~~](https://ews-build.webkit.org/#/builders/11/builds/37705 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/13540 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/14/builds/39440 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/39717 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/11796 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/14/builds/39440 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/40973 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/33639 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/14/builds/39440 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/40973 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/12122 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/51/builds/13185 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/40973 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/13750 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/14/builds/39440 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/12483 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4805 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/13014 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

psaavedra commented on Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp:
```diff
@@ -145,16 +150,20 @@ static std::unique_ptr<WPE::DRM::Connector> chooseConnector(int fd, drmModeRes*
-        if (resources->crtcs[i] == encoder->crtc_id) {
-            WPE::DRM::UniquePtr<drmModeCrtc> drmCrtc(drmModeGetCrtc(fd, resources->crtcs[i]));
-            return WPE::DRM::Crtc::create(fd, drmCrtc.get(), i);
+        for (int j = 0; j < resources->count_crtcs; ++j) {
```
Maybe we could still use the same code (the one remove by this patch) but also use the possible_crtcs check. Like @ceyusa  does here https://coral.googlesource.com/imx-gst-plugins-bad/+/refs/tags/1.14.4+imx-6/sys/kms/gstkmssink.c#241

carlosgcampos replied:
I don't understand what's wrong with this, or why this is not robust enough.

psaavedra replied:
The here point is that it can happen that the connector is not bound to an encoder.

In this scenario, the following check in `chooseCrtcForConnector()` is too restrictive because directly cuts the possibility of still finding a valid encoder for the connector:  
  
``` cpp
    if (!encoder) {
        g_set_error_literal(error, WPE_DISPLAY_ERROR, WPE_DISPLAY_ERROR_CONNECTION_FAILED, "Failed to get the connector encoder");
        return nullptr;
    }
```

This valid encoder it can by found by querying to the encoders if they can be a possible encoder for the connector:

``` cpp
    /* Try the currently connected encoder+crtc */
    for (int i = 0; i < resources->count_encoders; ++i) {

        WPE::DRM::UniquePtr<drmModeEncoder> encoder(drmModeGetEncoder(fd, resources->encoders[i]));
        if (!encoder)
            continue;

        for (int j = 0; j < resources->count_crtcs; ++j) {
            const uint32_t crtcMask = 1 << j;
            const uint32_t crtcId = resources->crtcs[j];
            if (encoder->possible_crtcs & crtcMask) {
                WPE::DRM::UniquePtr<drmModeCrtc> drmCrtc(drmModeGetCrtc(fd, crtcId));
                return WPE::DRM::Crtc::create(fd, drmCrtc.get(), j);
            }
        }
    }
```

psaavedra commented on Source/WebKit/WPEPlatform/wpe/drm/WPEDRM.cpp:
```diff
@@ -81,10 +81,6 @@ std::unique_ptr<Connector> Connector::create(int fd, drmModeConnector* connector
     if (!properties)
         return nullptr;
 
-    auto dpms = drmPropertyForName(fd, properties.get(), "DPMS");
```
This is a legacy property: https://drmdb.emersion.fr/properties/3233857728/DPMS

psaavedra replied:
I don't think there is any advantage with checking for this during the initialization.

carlosgcampos replied:
Fair enough, but let's split this, please use https://bugs.webkit.org/show_bug.cgi?id=268533 for this change only

psaavedra replied:
Implemented in https://github.com/WebKit/WebKit/pull/23654

psaavedra commented on Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp:
```diff
@@ -132,7 +137,7 @@ static std::unique_ptr<WPE::DRM::Connector> chooseConnector(int fd, drmModeRes*
 {
     for (int i = 0; i < resources->count_connectors; ++i) {
         WPE::DRM::UniquePtr<drmModeConnector> connector(drmModeGetConnector(fd, resources->connectors[i]));
-        if (!connector || connector->connection != DRM_MODE_CONNECTED || connector->connector_type == DRM_MODE_CONNECTOR_WRITEBACK || !connector->encoder_id)
```
This is the part where I don't know if I am doing the right thing of this is because a buggy result from the imx6 DRM driver. In my results I connect to the fd successfully but the connector returns a `encoder_id == 0` what seems a bit extrange. However I google some implementations (e.g: [1](https://gist.github.com/dvdhrm/5083553#file-modeset-c-L61) ) where they manage this case and they read this 0 like _the connector is not currently bound to an encoder or if the encoder+crtc is already used by another connector_.

psaavedra commented on Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp:
```diff
@@ -88,7 +88,12 @@ static UnixFileDescriptor findDevice(GError** error)
 
         fd = UnixFileDescriptor { open(device->nodes[DRM_NODE_PRIMARY], O_RDWR | O_CLOEXEC), UnixFileDescriptor::Adopt };
+
         if (fd)
```
There is a typo here, it should be `if (!fd)` .  ... Fixing the error.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp:
```diff
@@ -88,7 +88,12 @@ static UnixFileDescriptor findDevice(GError** error)
+            continue;
+
+        WPE::DRM::UniquePtr<drmModeRes> resources(drmModeGetResources(fd.value()));
+        if (resources)
```
You are continuing the loop when the fd is valid. I agree we could check that the device is actually a valid DRM device we can use for modesetting. Note that drmModeGetResources returns nullptr on error, not because the device doesn't have resources, I guess it's a permission error in you case. So, we could also check that indeed has resources with something like:

```
WPE::DRM::UniquePtr<drmModeRes> resources(drmModeGetResources(fd.value()));
if (resources && resources->count_crtcs && resources->count_connectors && resources->count_encoders)
```

But again, let's move this to its own PR, please use https://bugs.webkit.org/show_bug.cgi?id=268535

psaavedra replied:
I am seeing typos in this code (probably during the clean-up). I will separate the this in a single ticket as you suggest.

psaavedra replied:
Splitted  to: https://github.com/WebKit/WebKit/pull/23657

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp:
```diff
@@ -145,16 +145,30 @@ static std::unique_ptr<WPE::DRM::Connector> chooseConnector(int fd, drmModeRes*
 
 static std::unique_ptr<WPE::DRM::Crtc> chooseCrtcForConnector(int fd, drmModeRes* resources, const WPE::DRM::Connector& connector, GError** error)
 {
+    /* Try the currently connected encoder+crtc */
```
Use C++ style comments and finish with period.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp:
```diff
@@ -145,16 +145,30 @@ static std::unique_ptr<WPE::DRM::Connector> chooseConnector(int fd, drmModeRes*
-        if (resources->crtcs[i] == encoder->crtc_id) {
-            WPE::DRM::UniquePtr<drmModeCrtc> drmCrtc(drmModeGetCrtc(fd, resources->crtcs[i]));
-            return WPE::DRM::Crtc::create(fd, drmCrtc.get(), i);
+    /* If no active crtc was found, pick the first possible crtc, then try the first possible for crtc */
```
Ditto.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp:
```diff
@@ -145,16 +145,30 @@ static std::unique_ptr<WPE::DRM::Connector> chooseConnector(int fd, drmModeRes*
+
+        for (int j = 0; j < resources->count_crtcs; ++j) {
+            const uint32_t crtcMask = 1 << j;
+            const uint32_t crtcId = resources->crtcs[j];
```
Move this inside the if, or even better remove it and just use resources->crtcs[j].

psaavedra committed:
[WPE] Fix DRM initialization in nxp-imx6qp-sdb board
https://bugs.webkit.org/show_bug.cgi?id=268487

Reviewed by Carlos Garcia Campos.

This patch is generic but fixes the DRM initialization for NXP imx6qp
Sabre Development boards:

* Removes the check for the valid connector in the chooseConnector().
  The encoder Id can be zero if the connector is not currently bound
  to an encoder or if the encoder+crtc is already used by another
  connector.

* Modifies the chooseCrtcForConnector() logic. We implement a more
  robust mechanism by checking the encoder->possible_crtcs in case of
  no active crtc was found.

Credits to Jani Hautakangas <jani@igalia.com>.

Related-To: https://github.com/Igalia/cog/pull/590
Related-To: https://github.com/WebKit/WebKit/pull/23654
Related-To: https://github.com/WebKit/WebKit/pull/23657

* Source/WebKit/WPEPlatform/wpe/drm/WPEDisplayDRM.cpp:
(chooseConnector):
(chooseCrtcForConnector):

Canonical link: https://commits.webkit.org/273898@main
